✨ Add support for extending typer apps without passing a name, add commands to the top level (#1037)
fastapi/typer: docs/tutorial/splitting-apps.md

Reviewer: Let's make the name of the file (and URL) consistent with the title of the file.

Maybe naming it `multi-file-applications.md`, or maybe renaming the section to `# Splitting Apps`.

---

Actually, how about `# One File per Command` and `one-file-per-command.md`?

I guess we estimate that is the main use case, right?

It would be a bit more correct to have something like `# Include Commands from Apps` or `# Include Commands without Sub Commands`. But that's probably not as intuitive and obvious. :thinking: 

Not sure. Do you have any thoughts?

Author: Let's go with one file per command 😊

# Multi-File Applications

When your CLI application grows, you can split it into multiple files and modules. This pattern helps maintain clean and organized code structure.

This tutorial will show you how to use `add_typer` to create sub commands and organize your commands in multiple files.

We will create a simple CLI with the following commands:

- `version`
- `users add NAME`
- `users delete NAME`

## CLI structure

Here is the structure we'll be working with:

```text
mycli/
├── __init__.py
├── main.py
├── users/
│   ├── __init__.py
│   ├── add.py
│   └── delete.py
└── version.py
```

> `mycli` will be our package, and it will contain the following modules:

Reviewer: ```suggestion
`mycli` will be our <abbr title="a directory with an __init__.py file">package</abbr>, and it will contain the following modules:
```

> - `main.py`: The main module that will import the `version` and `users` modules.

Reviewer: ```suggestion
- `main.py`: The main <abbr title="a Python file that can be imported">module</abbr> that will import the `version` and `users` modules.
```

> - `version.py`: A module that will contain the `version` command.

Reviewer: ```suggestion
- `version.py`: A <abbr title="a Python file that can be imported">module</abbr> that will contain the `version` command.
```

> - `users/`: A package that will contain the `add` and `delete` commands.

Reviewer: ```suggestion
- `users/`: A <abbr title="another directory with an __init__.py file">package</abbr> (inside of our `mycli` package) that will contain the `add` and `delete` commands.
```

## Implementation

Let's start implementing our CLI! We'll create the `version` module, the `main` module, and the `users` package.

### Version Module (`version.py`)

Let's start by creating the `version` module. This module will contain the `version` command.

{* docs_src/splitting_apps/version.py *}

In this file we are creating a new Typer app instance for the version command. This is not required in single-file applications, but in the case of multi-file applications it will allow us to include this command in the main application using `add_typer`.

Let's see that next!

### Main Module (`main.py`)

> The main module will be the entry point of the application. It will import the version module and the users module. We'll see how to implement the user module in the next section.

Reviewer: ```suggestion
The main module will be the entry point of the application. It will import the version module and the users module.

/// tip

We'll see how to implement the user module in the next section.

///
```

Let's now create the `users` module with the `add` and `delete` commands.

### Users Add Command (`users/add.py`)

{* docs_src/splitting_apps/users/add.py *}

Like the `version` module, we create a new Typer app instance for the `users/add` command. This allows us to include the `add` command in the users app.

### Users Delete Command (`users/delete.py`)

{* docs_src/splitting_apps/users/delete.py *}

And once again, we create a new Typer app instance for the `users/delete` command. This allows us to include the `delete` command in the users app.

### Users' app (`users/__init__.py`)

Finally, we need to create an `__init__.py` file in the `users` directory to define the `users` app.

{* docs_src/splitting_apps/users/__init__.py *}

Similarly to the `version` module, we create a new `Typer` app instance for the `users` module. This allows us to include the `add` and `delete` commands in the users app.

## Running the Application

Now we are ready to run the application!

> To run the application, execute the `main.py` file:

Reviewer: ```suggestion
To run the application, you can execute it as a Python module:
```

<div class="termy">

```console

> $ python main.py version

Reviewer: ```suggestion
$ python -m mycli.main version
```

I followed the tutorial and copied the files, to make it work I had to update this.

My CLI Version 1.0

> $ python main.py users add Camila

Reviewer: ```suggestion
$ python -m mycli.main users add Camila
```

Adding user: Camila
```

> </div>

Reviewer: ```suggestion
</div>

And if you built a package and installed your app, you can then use the `mycli` command:

<div class="termy">

```console
$ mycli version

My CLI Version 1.0

$ mycli users add Camila

Adding user: Camila
```

</div>
```
